src/grpc/config/hpr_cs_skf_stream_worker.erl

```diff
@@ -70,65 +68,42 @@ handle_cast(Msg, State) ->
-            {noreply, State#state{conn_backoff = Backoff1}};
-        Connection ->
-            {_, Backoff1} = backoff:succeed(Backoff0),
-            #{http_connection := Pid} = Connection,
-            _Ref = erlang:monitor(process, Pid, [{tag, {'DOWN', ?MODULE}}]),
-            lager:info("connected"),
-            {ok, Stream} = grpc_client:new_stream(
-                Connection, 'helium.config.session_key_filter', stream, client_config_pb
-            ),
-            %% Sending SKF Stream Request
-            {PubKey, SigFun} = persistent_term:get(?HPR_KEY),
-            PubKeyBin = libp2p_crypto:pubkey_to_bin(PubKey),
-            SKFStreamReq = hpr_skf_stream_req:new(PubKeyBin),
-            SignedSKFStreamReq = hpr_skf_stream_req:sign(SKFStreamReq, SigFun),
-            ok = grpc_client:send_last(
-                Stream, hpr_skf_stream_req:to_map(SignedSKFStreamReq)
-            ),
-            lager:info("stream initialized"),
-            self() ! ?RCV_CFG_UPDATE,
-            {noreply, State#state{stream = Stream, conn_backoff = Backoff1}}
+            {noreply, State#state{conn_backoff = Backoff1}}
     end;
+%% GRPC stream callbacks
+handle_info({data, _StreamID, SKFStreamRes}, State) ->
+    lager:debug("sfk update"),
+    ok = process_res(hpr_skf_stream_res:from_map(SKFStreamRes)),
+    {noreply, State};
 handle_info(
-    ?RCV_CFG_UPDATE,
-    #state{
-        stream = Stream, conn_backoff = Backoff0
-    } = State
+    {'DOWN', Ref, process, Pid, _Reason},
+    #state{stream = #{stream_pid := Pid, monitor_ref := Ref}} = State
 ) ->
-    case grpc_client:rcv(Stream, ?RCV_TIMEOUT) of
-        {headers, _Headers} ->
-            self() ! ?RCV_CFG_UPDATE,
-            {noreply, State};
-        {data, SKFStreamRes} ->
-            ok = process_res(hpr_skf_stream_res:from_map(SKFStreamRes)),
-            lager:info("got sfk update"),
-            self() ! ?RCV_CFG_UPDATE,
-            {noreply, State};
-        eof ->
-            lager:warning("got eof"),
-            self() ! ?INIT_STREAM,
-            {noreply, State#state{stream = undefined}};
-        {error, timeout} ->
-            lager:debug("rcv timeout"),
-            {_, Backoff1} = backoff:succeed(Backoff0),
-            self() ! ?RCV_CFG_UPDATE,
-            {noreply, State#state{conn_backoff = Backoff1}};
-        {error, E} ->
-            lager:error("failed to rcv ~p", [E]),
-            self() ! ?INIT_STREAM,
-            {noreply, State#state{stream = undefined}}
-    end;
-handle_info({{'DOWN', ?MODULE}, _Mon, process, _Pid, _ExitReason}, State) ->
-    lager:info("connection ~p went down ~p", [_Pid, _ExitReason]),
+    lager:warning("stream closed"),
     self() ! ?INIT_STREAM,
     {noreply, State#state{stream = undefined}};
+handle_info({headers, _StreamID, _Headers}, State) ->
+    %% noop on headers
+    {noreply, State};
+handle_info({trailers, _StreamID, _Trailers}, State) ->
+    %% noop on trailers
+    {noreply, State};
```

Comment:
Should we do any stream cleanup on terminate?

Reply:
Yes, looking into it right now.

Reply:
Stream Pids are linked to their Channels. 
When the channel goes down, it takes down all it's streams.

Workers like this one clean up by dumping the `grpcbox_stream:stream()` map.

In the protocol router, when a stream receives an EOS, it removes itself from ets.